pagopa/io-app
# chore: [LLK-63] LolliPoP login code refactored to share common functionalities

⚠ This PR depends on #4443 ⚠

## Short description
This PR refactors the lolliPoP login code to share common functionalities, instead of duplicating them between `IdpLoginScreen` and `CieRequestAuthenticationOverlay`

## List of changes proposed in this pull request
- `ts/features/lollipop/store/reducers/lollipop.ts`: _redux in-memory store_ type edited to use `Option` from _fp-ts_ instead of optionals from _typescript_ (for both `keyTag` and `publicKey`) - state type, reducer and selectors;
- `ts/features/lollipop/store/index.ts`: added store version and migration for _redux-persist_, in order to migrate the persisted `keyTag` from `string?` to `Option<string>`;
- `ts/features/lollipop/hooks/useLollipopLoginSource.tsx`: shared code between `IdpLoginScreen` and `CieRequestAuthenticationOverlay` has been moved here. Removed the `LoginSource` data type since it was redundant;
- `ts/screens/authentication/IdpLoginScreen.tsx`: refactored in order to directly use shared code in `useLollipopLoginSource`;
- `ts/components/cie/CieRequestAuthenticationOverlay.tsx`: refactored in order to directly use shared code in `useLollipopLoginSource`;

## How to test
- Try the lolliPoP persisted store migration: make sure to have a previous application version installed, with a logged lolliPoP user. Update the application version to the one in this PR and make sure that the lolliPoP persisted store is migrated properly (both `keyTag` and `publicKey` have been wrapped in an `Option` and internal values have remained the same)
- Try the login flow, both for Spid and for CIE login, both on dev-server and on production and check that everything works fine. 
- Try the various failure flows and check that they are properly handled (either with a blocking screen or with a retry capability that effectively works)
